### PR TITLE
Use NVTX from GitHub

### PR DESCRIPTION
**Pull request by jrhemstad**
closes https://github.com/rapidsai/cudf/issues/6476

Adds CPM and uses it to fetch NVTX.

This required bumping the cmake minimum version to 3.18 to get support for `SOURCE_SUBDIR`. 

This currently depends on my fork of `NVTX` until https://github.com/NVIDIA/NVTX/pull/14 is merged.

Depends on https://github.com/rapidsai/integration/pull/156

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -788,11 +790,23 @@ message(STATUS "RMM_LOGGING_LEVEL = '${RMM_LOGGING_LEVEL}'.")
 
 target_compile_definitions(cudf PUBLIC SPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_${RMM_LOGGING_LEVEL})
 
+###################################################################################################
+# - nvtx ------------------------------------------------------------------------------------------
+CPMAddPackage(
```
Comment (harrism) [inline cpp/CMakeLists.txt:795]: I think this supports the GIT_SHALLOW option?

Comment (jrhemstad) [reply cpp/CMakeLists.txt:795]: done

File: cpp/CMakeLists.txt [hunk 2]
```diff
@@ -13,14 +13,16 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #=============================================================================
-cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
```
Comment (kkraus14) [inline cpp/CMakeLists.txt:16]: If we're bumping the minimum to 3.18 we should probably set policy CMP0104 to `OLD` until we use the proper `CMAKE_CUDA_ARCHITECTURES` / `CUDA_ARCHITECTURES` options.

https://cmake.org/cmake/help/v3.18/policy/CMP0104.html

File: cpp/cmake/CPM.cmake [hunk 3]
```diff
@@ -0,0 +1,19 @@
+set(CPM_DOWNLOAD_VERSION c3c7e2d9a3b890511891ffc415d7ea23c689068a) # 0.27.3
```
Comment (kkraus14) [inline cpp/cmake/CPM.cmake:1]: We should actually bump to 0.27.4 as there was some nice fixes related to specifying a source directory via cmake variables and being able to pass options through.

File: cpp/CMakeLists.txt (hunk 2 above)
Comment (jrhemstad) [reply cpp/CMakeLists.txt:16]: I'm not sure what this means

Comment (kkraus14) [reply cpp/CMakeLists.txt:16]: `CMP0104` is a CMake policy that was added in CMake 3.18 that automatically sets the `CUDA_ARCHITECTURES` property of all targets to `CMAKE_CUDA_ARCHITECTURES` which if not set by the user, defaults to the oldest supported architecture of nvcc (Maxwell, 52 for CUDA 11.0).

By setting the policy to `OLD` it reverts to the pre CMake 3.18 behavior where `CUDA_ARCHITECTURES` is not implicitly set, so we can continue to use the nvcc compiler flags as we're currently doing.

Comment (jrhemstad) [reply cpp/CMakeLists.txt:16]: I'm not familiar with cmake policies. How do I set it to OLD?

Comment (kkraus14) [reply cpp/CMakeLists.txt:16]: `cmake_policy(CMP0104 OLD)`

**Comment by jrhemstad** [comment]
rerun tests

**Comment by kkraus14** [comment]
FYI the gpu tests will fail unless you temporarily do something like https://github.com/rapidsai/rmm/pull/604/files#diff-285f239638b8b5c5499ef5fda2e3db69adda7045e81556b4f2df478e99ab6263R51-R52 to bump cmake to 3.18.

**Comment by harrism** [comment]
@jrhemstad should we push this to 0.18?

**Comment by harrism** [comment]
Move to 0.18

**Comment by harrism** [comment]
@jrhemstad I missed this because it lacked the libcudf label. Plans for it?

**Comment by jrhemstad** [comment]
Moving to 21.08...

**Comment by bdice** [comment]
Closing as stale - this PR would need to be redone with rapids-cmake in order to resolve #6476, and it seems we may still be waiting on a new release/tag of NVTX upstream.

**Pull request closed without merging**